game-core/src/main/java/org/triplea/thread/LockUtil.java

```diff
@@ -1,154 +0,0 @@
-package org.triplea.thread;
```

Comment:
I assume you mean "is replaced" instead of "is not replaced".

Almost. The only usage that remained after using standard synchronisation as a replacement is the 4 usages in GameData, where I just called lock and unlock directly. Because we're never really dealing with multiple GameDatas at the same time the read and write locks essentially become the only locks that are used via the LockUtil class, which defeats its purpose a little bit. After all it's supposed to keep track of the locks acquired and released by itself.

Not quite that satisfying, but I guess it's a start.

Reply:
> Because we're never really dealing with multiple GameDatas at the same time the read and write locks essentially

How do you mean? There are many copies of game data taken at several different points, some notable examples: battle-calc, history, and I presume even network.

Reply:
True, but I don't think there's any moment where we're holding write/read locks of 2 different instances at the same time.
I might be wrong though, so feel free to correct me

Reply:
My impression from surveying the locks, there are only a few write locks, but many read locks. Read locks are grabbed hundreds of times per second from all over the place. For example, I did some debugging in lock util and added a 20ms - 200ms wait when acquiring a read lock, and the game never launched it was so busy doing busy waits. I would first assume there is concurrency between read locks unless able to prove otherwise.

Reply:
Yeah, but this particular class only had a single non-static re-entrant lock and now it has a single non-static object everything is synchronised on.
The only class working with readwrite locks here is GameData and the Tile class has nothing to do with that.

Reply:
AFAIK, the locks that are shared are through game data, which in turn is where all the locking is done for the rest of the system (back to game data).

I suspect the perf is probably better now actually. `LockUtil` had a couple methods where if you repeated the 1000 times (I was trying to simulate a deadlock on them not too long ago), the test starts to get really slow and each iteration slows down to be a second and more. The `weakref` mechanism seemed to have been the culprit in that case, large set copies were happening. Having that removed, we've a simpler system which is more likely to be fast and/or easily made fast.